src/GraphQL/Types/Scalars/ComplexScalarGraphType.cs

```diff
@@ -0,0 +1,118 @@
+    {
+        Name = "Complex";
+    }
+
+    /// <inheritdoc/>
+    public override object? ParseLiteral(GraphQLValue value) => value.ParseAnyLiteral();
+
+    /// <inheritdoc/>
+    public override object? ParseValue(object? value) => value;
+
+    /// <inheritdoc/>
+    public override bool CanParseLiteral(GraphQLValue value) => value != null;
+
+    /// <inheritdoc/>
+    public override bool CanParseValue(object? value) => true;
+
+    /// <inheritdoc/>
+    public override object? Serialize(object? value) => value;
+
+    /// <inheritdoc/>
+    public override GraphQLValue ToAST(object? value)
+    {
+        return value switch
+        {
+            null => GraphQLValuesCache.Null,
+
+            true => GraphQLValuesCache.True,
+            false => GraphQLValuesCache.False,
+
+            byte n => new GraphQLIntValue(n),
+            sbyte n => new GraphQLIntValue(n),
+            short n => new GraphQLIntValue(n),
+            ushort n => new GraphQLIntValue(n),
+            int n => new GraphQLIntValue(n),
+            uint n => new GraphQLIntValue(n),
+            long n => new GraphQLIntValue(n),
+            ulong n => new GraphQLIntValue(n),
+            BigInteger n => new GraphQLIntValue(n),
+
+            float n => new GraphQLFloatValue(n),
+            double n => new GraphQLFloatValue(n),
+            decimal n => new GraphQLFloatValue(n),
+
+            string s => new GraphQLStringValue(s),
+
+            IDictionary d => ConvertDictionary(d),
+            IEnumerable l => ConvertList(l),
+
+            _ => throw new NotSupportedException($"Converting the complex type '{value.GetType().GetFriendlyName()}' to its AST representation is not supported.")
+        };
+
+        GraphQLObjectValue ConvertDictionary(IDictionary dictionary)
+        {
+            if (dictionary.Count == 0)
+                return new();
+
+            if (dictionary is IDictionary<string, object?> objectDictionary)
+            {
+                return new()
+                {
+                    Fields = objectDictionary
+                        .Select(row => new GraphQLObjectField(
+                            new GraphQLName(row.Key),
+                            ToAST(row.Value)))
+                        .ToList(),
+                };
+            }
+
+            var fields = new List<GraphQLObjectField>(dictionary.Count);
+            var enumerator = dictionary.GetEnumerator();
+            while (enumerator.MoveNext())
+            {
+                if (enumerator.Key is not string keyString)
+                    throw new InvalidOperationException("Object keys must be string values");
+                fields.Add(new GraphQLObjectField(
+                    new GraphQLName(keyString),
+                    ToAST(enumerator.Value)));
+            }
+            return new() { Fields = fields };
+        }
+
+        GraphQLListValue ConvertList(IEnumerable list)
+        {
+            List<GraphQLValue>? values = null;
```

Comment:
It is worth checking if the `list is ICollection` and allocating the list with the initial capacity to avoid re-sizes

Reply:
```suggestion
            List<GraphQLValue>? values = null;
            if (list is ICollection collection)
                values = new(collection.Count);
```

Reply:
Good idea